components/core/src/os/net/mod.rs

```diff
@@ -22,3 +22,73 @@ mod imp;
 mod imp;
 
 pub use self::imp::*;
+
+extern crate dns_lookup;
+
+use dns_lookup::{getaddrinfo,
+                 AddrInfoHints};
+use std::io;
+
+// lookup_fqdn returns a vector of fqdn that resolves the provided hostname.
+//
+// Since the underlying crate is platform agnostic, this function is as well,
+// we only implement a single variable `flags` that has the right hint for
+// the running operating system.
+pub fn lookup_fqdn(hostname: &str) -> io::Result<String> {
+    #[cfg(not(windows))]
+    let flags = libc::AI_CANONNAME;
+
+    #[cfg(windows)]
+    let flags = winapi::shared::ws2def::AI_CANONNAME;
+
+    let hints = AddrInfoHints { flags,
+                                ..AddrInfoHints::default() };
+    let addrinfos =
+        getaddrinfo(Some(hostname), None, Some(hints))?.collect::<std::io::Result<Vec<_>>>()?;
+
+    // If 'hints.ai_flags' includes the AI_CANONNAME flag, then the ai_canonname
+    // field of the first of the addrinfo structures in the returned list is set
+    // to point to the official name of the host.
+    if !addrinfos.is_empty() {
+        let addrinfo = addrinfos[0].clone();
+        return Ok(addrinfo.canonname.unwrap_or_else(|| hostname.to_string()));
+    }
+
+    Ok(hostname.to_string())
+}
+
+// fqdn returns the fully qualified domain name of the running machine
+pub fn fqdn() -> Option<String> {
+    match hostname() {
+        Ok(hostname) => lookup_fqdn(&hostname).ok(),
+        Err(_) => None,
+    }
```

Comment:
I'd be interested to see what the error cases here are (both in the `ok()` case, as well as the `Err(_)` case).

Logging those at a debug level would be nice.

(I'm also curious as to the various failure modes involved here, and what they imply.)

Reply:
Good idea! Added the debug statements 💯

Reply:
We can factor out the redundancy in the debug statements here with [`and_then`](http://doc.rust-lang.org/1.34.1/std/result/enum.Result.html#method.and_then):
```rust
    match dns_lookup::get_hostname().and_then(|hostname| lookup_fqdn(&hostname)) {
        Ok(fqdn) => Some(fqdn),
        Err(e) => {
            debug!("fqdn() was unable to lookup the machine fqdn. {:?}", e);
            None
        }
    }
```
Or, even a bit more concisely if we convert the `Result` -> `Option` using [`Result::ok`](http://doc.rust-lang.org/1.34.1/std/result/enum.Result.html#method.ok):
```rust
    let result = dns_lookup::get_hostname().and_then(|hostname| lookup_fqdn(&hostname));
    if let Err(ref e) = result {
        debug!("fqdn() was unable to lookup the machine fqdn. {:?}", e);
    }
    result.ok()
```

Reply:
I used the super concisely way (the second one). Though I have to say that even though my implementation was a bit more code (a bit more ugly 😅 ) it was a bit more helpful since you could have known whether this function was unable to get the machine `hostname` or get the machine `fqdn`, at least more clear and helpful for the end-user or developers that are troubleshooting anything at this level of the code base.

Reply:
If you think that distinction will be helpful, you don't *need* to make this change. It was a suggestion of how I'd have done it.

Looking at http://man7.org/linux/man-pages/man2/gethostname.2.html, it seems pretty unlikely that `get_hostname` will fail, and I tend towards simplicity and non-redundancy in the first revision since complexity is harder to reduce later. If we discover it's useful to log these errors distinctly, we can add it later (though that first time may be painful).

In any case, you're the author here and I want you to do what feels best to you. The fact that you made this comment tells me you're not entirely happy with my suggestion. If I see something that I feel is very important to change, I'll make that clear. Otherwise, you can interpret my comments as suggestions and alternative perspectives, not mandates.

Reply:
This was absolutely a misunderstanding from my side Jon, I took your feedback very seriously and I shouldn't have done that. I promise to do what you said in your comment above next time and take it as a suggestion. My apologies.

Reply:
Though, it is almost impossible for me not to change my code with your suggestions since they are so badass!!! Thank you so much for all your feedback. ❤️

Reply:
No apologies necessary! I'm happy to give lots of feedback. It helps me learn the new code and everyone gets better though seeing more different approaches and perspectives.

It's awesome to have another new rustacean!